src/lib/hooks/swap/useSwapInfo.tsx

```diff
@@ -0,0 +1,157 @@
+  } = useAtomValue(swapAtom)
+
+  const to = account
+
+  const relevantTokenBalances = useCurrencyBalances(
+    account,
+    useMemo(() => [inputCurrency ?? undefined, outputCurrency ?? undefined], [inputCurrency, outputCurrency])
+  )
+
+  const isExactIn: boolean = independentField === Field.INPUT
+  const parsedAmount = useMemo(
+    () => tryParseCurrencyAmount(amount, (isExactIn ? inputCurrency : outputCurrency) ?? undefined),
+    [inputCurrency, isExactIn, outputCurrency, amount]
+  )
+
+  const trade = useBestTrade(
+    isExactIn ? TradeType.EXACT_INPUT : TradeType.EXACT_OUTPUT,
+    parsedAmount,
+    (isExactIn ? outputCurrency : inputCurrency) ?? undefined
+  )
+
+  const currencies = useMemo(
+    () => ({
+      [Field.INPUT]: inputCurrency ?? undefined,
+      [Field.OUTPUT]: outputCurrency ?? undefined,
+    }),
+    [inputCurrency, outputCurrency]
+  )
+
+  const currencyBalances = useMemo(
+    () => ({
+      [Field.INPUT]: relevantTokenBalances[0],
+      [Field.OUTPUT]: relevantTokenBalances[1],
+    }),
+    [relevantTokenBalances]
+  )
+
+  const currencyAmounts = useMemo(
+    () => ({
+      [Field.INPUT]: trade.trade?.inputAmount,
+      [Field.OUTPUT]: trade.trade?.outputAmount,
+    }),
+    [trade.trade?.inputAmount, trade.trade?.outputAmount]
+  )
+
+  // TODO(ianlapham): Fix swap slippage tolerance
+  // const allowedSlippage = useSwapSlippageTolerance(trade.trade ?? undefined)
+  const allowedSlippage = useMemo(() => new Percent(100), [])
+
+  const inputError = useMemo(() => {
+    let inputError: ReactNode | undefined
+
+    if (!account) {
+      inputError = <Trans>Connect Wallet</Trans>
+    }
+
+    if (!currencies[Field.INPUT] || !currencies[Field.OUTPUT]) {
+      inputError = inputError ?? <Trans>Select a token</Trans>
+    }
+
+    if (!parsedAmount) {
+      inputError = inputError ?? <Trans>Enter an amount</Trans>
+    }
+
+    const formattedTo = isAddress(to)
+    if (!to || !formattedTo) {
+      inputError = inputError ?? <Trans>Enter a recipient</Trans>
+    } else {
+      if (BAD_RECIPIENT_ADDRESSES[formattedTo]) {
+        inputError = inputError ?? <Trans>Invalid recipient</Trans>
+      }
+    }
+
+    // compare input balance to max input based on version
+    const [balanceIn, amountIn] = [currencyBalances[Field.INPUT], trade.trade?.maximumAmountIn(allowedSlippage)]
+
+    if (balanceIn && amountIn && balanceIn.lessThan(amountIn)) {
+      inputError = <Trans>Insufficient {amountIn.currency.symbol} balance</Trans>
+    }
+
+    return inputError
+  }, [account, allowedSlippage, currencies, currencyBalances, parsedAmount, to, trade.trade])
+
+  return useMemo(
+    () => ({
+      currencies,
+      currencyBalances,
+      currencyAmounts,
+      inputError,
+      trade,
+      allowedSlippage,
+    }),
+    [currencies, currencyBalances, currencyAmounts, inputError, trade, allowedSlippage]
+  )
+}
+
+const swapInfoAtom = atom<SwapInfo>({
+  currencies: {},
```

Comment:
I think `currencies` should be removed to encourage fetching them through `Swap` state. It's available through `currencyAmounts`, in case you want to fetch everything in one go. wdyt?

Reply:
I think the issue with only relying on currencyAmounts, you dont get access to the currency until an amount has been populate 

a related issue to the setup of this state is that when both currencies are selected - theres a lag in typing - might be related to some of teh changes we made

Reply:
I think the lag has to do with cosmos more than with our own state structures.